### PR TITLE
Add Arbor backend to GitHub Actions testing matrix

### LINKED ISSUE #1556: Add Arbor test to Travis
Currently, the Arbor-code in NEST goes completely untested and is not even compiled. #1535 disables it by default. The Travis test matrix should be enhanced to cover Arbor testing, at least a compile of the source code.

### PR DESCRIPTION
**Pull request by hakonsbm**
Also adds a test of the Arbor backend, based on the Arbor cosim example. Fixes #1556.

### CODE REVIEW

**Comment by lekshmideepu** [comment]
I would take a look!

File: pynest/nest/tests/arbor_backend_arbor_side.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+# This script simulates an Arbor run with MPI spike exchange to external
+# NEST. It saves the received spikes to file for testing.
+# Based on arbor_proxy.py in arbor_cosim_example.
+from mpi4py import MPI
```
Comment (gtrensch) [inline pynest/nest/tests/arbor_backend_arbor_side.py:25]:
```suggestion

from mpi4py import MPI
```

Comment (gtrensch) [reply pynest/nest/tests/arbor_backend_arbor_side.py:25]: Inserted a line break.

**Review by gtrensch — CHANGES_REQUESTED** [review]
That looks pretty good to me so far. 
Aside from the failed build due to mpi4py on macOS that needs to be investigated by a macOS expert, I made a few small suggestions to improve the comments. Could you check this please?

File: pynest/nest/tests/arbor_backend_arbor_side.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+# This script simulates an Arbor run with MPI spike exchange to external
+# NEST. It saves the received spikes to file for testing.
+# Based on arbor_proxy.py in arbor_cosim_example.
```
Comment (heplesser) [inline pynest/nest/tests/arbor_backend_arbor_side.py:24]: Change file name to `arbor_backend_test_arbor_side.py` and add a comment here that this file is run by `test_arbor_backend.py` together with `arbor_backend_test_nest_side.py`.

File: pynest/nest/tests/arbor_backend_nest_side.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+# This script runs NEST and transmits spikes via the Arbor backend.
+# Spikes are also stored to file for testing.
+# It requires NESTIO + ARBOR-NESTIO.
```
Comment (heplesser) [inline pynest/nest/tests/arbor_backend_nest_side.py:24]: Change file name to `arbor_backend_test_nest_side.py` and add a comment here that this file is run by `test_arbor_backend.py` together with `arbor_backend_test_arbor_side.py`.

File: pynest/nest/tests/test_arbor_backend.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# NEST is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
```
Comment (heplesser) [inline pynest/nest/tests/test_arbor_backend.py:20]: Add a comment here that this file requires `arbor_backend_test_arbor_side.py` and `arbor_backend_test_nest_side.py`.

File: pynest/nest/tests/test_arbor_backend.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        self.assertTrue(nest_output_file_exists)
+        self.assertTrue(arbor_output_file_exists)
+
+        nest_spikes = np.loadtxt(self.nest_output_fname, skiprows=3)
+        arbor_spikes = np.loadtxt(self.arbor_output_fname)
+        arbor_spikes[:, 0] -= 100  # Need to shift Arbor IDs to make them comparable to NEST IDs
+        np.testing.assert_array_equal(nest_spikes, arbor_spikes)
+
+
+def suite():
```
Comment (heplesser) [inline pynest/nest/tests/test_arbor_backend.py:80]: Since we now run via PyTest (see #2127), everything below is no longer needed.

**Review by heplesser — CHANGES_REQUESTED** [review]
Looks good, just some small comments below. Could you merge master to switch test running to pytest? That should also solve the problem on macOs.

**Review by heplesser — APPROVED** [review]

**Comment by heplesser** [comment]
@hakonsbm macOS still fails, even though mpi4py is definitely installed. I will take another look on the Mac.

File: .github/workflows/nestbuildmatrix.yml [hunk 6]
```diff
@@ -47,6 +50,9 @@ jobs:
            - xNEST_BUILD_TYPE: "OPENMP_ONLY"
              cpp_compiler: "gcc"
              os: macos-latest
+           - xNEST_BUILD_TYPE: "ARBOR_BACKEND"
+             cpp_compiler: "gcc"
+             os: macos-latest
```
Comment (heplesser) [inline .github/workflows/nestbuildmatrix.yml:55]: Insert below here
```
   - xNEST_BUILD_TYPE: "ARBOR_BACKEND"
              cpp_compiler: "clang"
              os: macos-latest
```
to completely prevent Arbor testing on macOS. At present, we only perform a minimal test on macOS to limit resource use; not all necessary requirements for a full test are installed.

**Review by heplesser — CHANGES_REQUESTED** [review]
I think the proper solution for now is to not test the Arbor backend on macOS; our test setup for macOS does not provide all necessary components. I suggested a change for that below.

Python-based tests have moved in master. Please move the Arbor tests from `pynest/nest/tests` to `testsuite/pytests`.

**Review by heplesser — APPROVED** [review]

**Review by gtrensch — APPROVED** [review]
👍 Approved.

**Pull request merged**